### PR TITLE
[SDC] Integrate deterministic replay into Trainer

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #4358
* #4357

## Summary

Add opt-in deterministic replay for detecting silent data corruption during
training.

The Trainer builds an SDCReplayer from config when the feature is enabled and
holds it as self.sdc_replayer (None when disabled). For each optimizer step,
the step's first forward/backward -- one full gradient-accumulation group,
or one complete pipeline schedule under pipeline parallelism -- is routed
through sdc_replayer.run_fwd_bwd, which replay-checks it when scheduled. A
mismatch raises SDCReplayMismatch on every rank before gradient clipping,
optimizer, scheduler, or checkpoint updates can observe corrupted state.

The replayer is built from config with narrow runtime inputs (modules,
device, and a ScalarStateAccessor for ntokens_seen, which the replayed
forward/backward mutates); it does not receive the Trainer. All replay
scheduling state lives inside the replayer; the check schedule is
process-local, not checkpointed, and restarts after checkpoint loading via
reset_schedule.

GraphTrainer shares the same replay boundary by inheriting Trainer.

## Configuration

The feature is disabled by default and has no CLI surface (sdc_replay is
suppressed from tyro parsing); enable it programmatically:

```python
config.debug.deterministic = True
config.sdc_replay.enabled = True
config.sdc_replay.num_steps = 1    # optimizer steps checked after each (re)start; -1 checks every step
config.sdc_replay.num_replays = 1  # re-executions compared against the reference
```

Replay requires debug.deterministic=True and
debug.deterministic_warn_only=False. Eager execution, torch.compile, CUDA
graphs, symmetric-memory FSDP, distributed GEMM, async TP, DeepEP v2,
HybridEP, and MinimalAsyncEP use the same replay boundary. CUDA graphs
require num_replays=1; this is validated purely from config (independent of
runtime CUDA graph availability) because additional replays would require
CUDA-graph-aware restoration of graph-owned gradient and optional-buffer
storage.

After config overrides are applied, Trainer re-runs the full
Config.__post_init__ (raise-only) instead of individual validators.

Document the feature under docs/observability/ and add CPU and GPU coverage
including an intentionally corrupted distributed gradient.

## Testing

- pytest -q tests/unit_tests/cpu/test_trainer.py
  tests/unit_tests/cpu/observability/test_sdc_replay.py
  tests/unit_tests/cpu/test_invalid_loss.py
  tests/unit_tests/cpu/test_config_manager.py
- tests/integration_tests/features.py::sdc_replay_cudagraph
- torchtitan/experiments/graph_trainer/tests/integration_tests.py::sdc_replay
- tests/integration_tests/h100.py::deepseek_v3_fsdp+cp+tp+minimal_async_ep+sdc_replay
- tests/integration_tests/features.py::sdc_replay_mismatch
- git diff --check
- pre-commit run --all-files

stack-info: PR: https://github.com/pytorch/torchtitan/pull/4334, branch: xmfan/stack/45

### CODE REVIEW

**Comment by xmfan** [comment]
accidentally used ghstack
